src/algebra/lie_algebra.lean

```diff
@@ -0,0 +1,214 @@
+-/
+class has_bracket (L : Type v) := (bracket : L → L → L)
+
+notation `⁅`x`,` y`⁆` := has_bracket.bracket x y
+
+namespace ring_commutator
+
+variables {A : Type v} [ring A]
+
+/--
+The ring commutator captures the extent to which a ring is commutative. It is identically zero
+exactly when the ring is commutative.
+-/
+def commutator (x y : A) := x*y - y*x
+
+local notation `⁅`x`,` y`⁆` := commutator x y
+
+@[simp] lemma add_left (x y z : A) :
+  ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆ :=
+by { unfold commutator, simp [right_distrib, left_distrib] }
+
+@[simp] lemma add_right (x y z : A) :
+  ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆ :=
+by { unfold commutator, simp [right_distrib, left_distrib] }
+
+@[simp] lemma alternate (x : A) :
+  ⁅x, x⁆ = 0 :=
+by { unfold commutator, simp }
+
+lemma jacobi (x y z : A) :
+  ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0 :=
+begin
+  unfold commutator,
+  repeat { rw mul_sub_left_distrib },
+  repeat { rw mul_sub_right_distrib },
+  repeat { rw add_sub },
+  repeat { rw ←sub_add },
+  repeat { rw ←mul_assoc },
+  have h : ∀ (x y z : A), x - y + z + y = x+z := by simp,
+  repeat { rw h },
+  simp,
+end
+
+end ring_commutator
+
+/--
+A Lie ring is an additive group with compatible product, known as the bracket, satisfying the
+Jacobi identity. The bracket is not associative unless it is identically zero.
+-/
+class lie_ring (L : Type v) [add_comm_group L] extends has_bracket L :=
+(add_lie : ∀ (x y z : L), ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆)
+(lie_add : ∀ (x y z : L), ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆)
+(lie_self : ∀ (x : L), ⁅x, x⁆ = 0)
+(jacobi : ∀ (x y z : L), ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0)
+
+section lie_ring
+
+variables {L : Type v} [add_comm_group L] [lie_ring L]
+
+@[simp] lemma add_lie (x y z : L) : ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆ := lie_ring.add_lie x y z
+@[simp] lemma lie_add (x y z : L) : ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆ := lie_ring.lie_add x y z
+@[simp] lemma lie_self (x : L) : ⁅x, x⁆ = 0 := lie_ring.lie_self x
+
+@[simp] lemma lie_skew (x y : L) :
+  -⁅y, x⁆ = ⁅x, y⁆ :=
+begin
+  symmetry,
+  rw [←sub_eq_zero_iff_eq, sub_neg_eq_add],
+  have H : ⁅x + y, x + y⁆ = 0 := by apply lie_self,
+  rw [add_lie, lie_add, lie_add, lie_self, lie_self] at H,
+  simp at H,
+  exact H,
+end
+
+@[simp] lemma lie_zero (x : L) :
+  ⁅x, 0⁆ = 0 :=
+begin
+  have H : ⁅x, 0⁆ + ⁅x, 0⁆ = ⁅x, 0⁆ + 0 := by { rw ←lie_add, simp, },
+  exact add_left_cancel H,
+end
+
+@[simp] lemma zero_lie (x : L) :
+  ⁅0, x⁆ = 0 := by { rw [←lie_skew, lie_zero], simp, }
+
+@[simp] lemma neg_lie (x y : L) :
+  ⁅-x, y⁆ = -⁅x, y⁆ := by { rw [←sub_eq_zero_iff_eq, sub_neg_eq_add, ←add_lie], simp, }
+
+@[simp] lemma lie_neg (x y : L) :
+  ⁅x, -y⁆ = -⁅x, y⁆ := by { rw [←lie_skew, ←lie_skew], simp, }
+
+@[simp] lemma gsmul_lie (x y : L) (n : ℤ) :
+  ⁅n • x, y⁆ = n • ⁅x, y⁆ :=
+begin
+  have H : is_add_group_hom (λ z, ⁅z, y⁆) := { map_add := by { intros, rw add_lie, } },
+  exact (@is_add_group_hom.map_gsmul _ _ _ _ _ H x n),
```

Comment:
You should probably read the documentation on instances management tactics at https://github.com/leanprover-community/mathlib/blob/master/docs/tactics.md#instance-cache-tactics. Then you'll be able to replace the above proof with:
```lean
  haveI : is_add_group_hom (λ z, ⁅z, y⁆) := { map_add := by { intros, rw add_lie, } },
  apply is_add_group_hom.map_gsmul (λ (z : L), ⁅z,y⁆)
```

Reply:
Marvellous! I've wanted to use `haveI` several times already, and resigned myself to it just not being a feature. This is a very welcome tip!